File: db-init/migrate.py
import re
import os

user_mapping = {
    'andrade': '1', 'arboilito': '2', 'chosina': '3', 'crepusculobra': '4',
    'danone': '5', 'jcamaney': '6', 'koreanite': '7', 'miniona': '8',
    'nemo12': '9', 'ponchalo': '10', 'rnata': '11'
}

hash_to_plain = {
    '$2b$10$n2TX9TtudzDP8lniHpFhXeoa/.nesXPtLwrjr5Zt30AAYNlyUo7KW': '1',
    '$2b$10$nBAFdUOMEK5litSyVEg/UOvcBFdCcxTxp9ctbNQ5eRazTeNDrdwGi': '1234',
    '$2b$10$JHWyY0UiBIN23YE7Vn8wkOqheV.iSCpvvX38HvYwSsZ7AFnyQun/S': '123',
    '$2b$10$vH/3cNMiAl5fpK8zjFfvPO6V1naNqO33TQ4.yJz5m.q7ZfSq5MgKC': '123',
    '$2b$10$cpRu/Lx561rVy6EQft2i7OW8bnPZLIG2xPYyPc6piFPBxK9gh07Py': '1234',
    '$2b$10$e3ltDmY4jX0UWGU.rr/HF.NCkZwqcVr1QjOAIpzpNC93iJ4/8ZzUS': '',
    '$2b$10$lyEQpSRipZBmN3OBNEPWa.e9h8ToyLRF4gSXq3qVWYyZ7a5JMiBBy': 'miamors',
    '$2b$10$PX9g.GBUo6RoOb/EwuLnOO6kvtPt8yMhsO.AgR66V8xNe1EjgMOPC': '1234',
    '$2b$10$sFcQ4YGOaIGWio9XRYWD4eqdZ/h4b9CkIpawcsHzp0xun6EXWyOoa': 'renata',
    '$2b$10$bE.ceZevG5jcHv09fXXPxOSGkT6lN7Cq1f7yY0QXnil6pwSDIEmHm': '123',
    '$2b$10$EG5FCjIwX6eqgMlsQ.Tzie47IaPTyIqA8coX4cOqp2yxO0H6qfmxe': 'llanas'
}

def process_file(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    # 1. Remove contrasenas
    pattern_contrasenas = re.compile(r'DROP TABLE IF EXISTS `contrasenas`;.*?UNLOCK TABLES;\s*', re.DOTALL)
    content = pattern_contrasenas.sub('', content)

    # 2. Modify Trabajadores schema
    old_schema = re.compile(r'CREATE TABLE `Trabajadores` \([^)]*PRIMARY KEY \(`usuario`\)\s*\) ENGINE=InnoDB.*?;', re.DOTALL)
    new_schema = """CREATE TABLE `Trabajadores` (
  `id` int NOT NULL AUTO_INCREMENT,
  `usuario` varchar(20) NOT NULL,
  `contrasena` varchar(128) NOT NULL,
  `texto_plano` varchar(25) DEFAULT NULL,
  `rol` varchar(15) NOT NULL,
  `Nombre` varchar(30) NOT NULL,
  `apellido_paterno` varchar(20) NOT NULL,
  `apellido_materno` varchar(20) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `usuario` (`usuario`)
) ENGINE=InnoDB AUTO_INCREMENT=20 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;"""
    content = old_schema.sub(new_schema, content)

    # 3. Process Trabajadores Inserts
    def replacer_trabajadores(m):
        raw_vals = m.group(1)
        # Split by format ),( or ), (
        parts = re.split(r'\),\s*\(', raw_vals.strip('()'))
        new_parts = []
        for p in parts:
            p = p.strip("()")
            # Elements e.g. 'andrade','$2b$10$xxx','Operario','Andrade','Chehue','Hernández'
            elems = re.findall(r"'[^']*'", p)
            if len(elems) == 6:
                user = elems[0].strip("'")
                hash_val = elems[1].strip("'")
                uid = user_mapping.get(user, '99')
                plain = hash_to_plain.get(hash_val, '')
                new_p = f"{uid},{elems[0]},{elems[1]},'{plain}',{elems[2]},{elems[3]},{elems[4]},{elems[5]}"
                new_parts.append(f"({new_p})")
            else:
                new_parts.append(f"({p})")
        return "INSERT INTO `Trabajadores` VALUES " + ",".join(new_parts) + ";"

    content = re.sub(r'INSERT INTO `Trabajadores` VALUES (.*?);', replacer_trabajadores, content)

    # 4. Modify Ventas table logic
    content = re.sub(r'`usuario` varchar\(20\) DEFAULT NULL,', r'`trabajador_id` int DEFAULT NULL,', content)
    content = re.sub(r'KEY `usuario` \(`usuario`\),', r'KEY `trabajador_id` (`trabajador_id`),', content)
    content = re.sub(r'CONSTRAINT `ventas_ibfk_2` FOREIGN KEY \(`usuario`\) REFERENCES `trabajadores` \(`usuario`\)', 
                     r'CONSTRAINT `ventas_ibfk_2` FOREIGN KEY (`trabajador_id`) REFERENCES `Trabajadores` (`id`)', content, flags=re.IGNORECASE)

    # 5. Process Ventas Inserts
    def replacer_ventas(m):
        raw_vals = m.group(1)
        parts = re.split(r'\),\s*\(', raw_vals.strip('()'))
        new_parts = []
        for p in parts:
            p = p.strip("()")
            # split by comma, ignoring commas inside quotes
            # But the last value is always the user string e.g. 'danone'
            elems = p.split(',')
            user = elems[-1].strip("'")
            if user in user_mapping:
                elems[-1] = user_mapping[user]
            new_parts.append("(" + ",".join(elems) + ")")
        return "INSERT INTO `ventas` VALUES " + ",".join(new_parts) + ";"

    content = re.sub(r'INSERT INTO `ventas` VALUES (.*?);', replacer_ventas, content)

    with open(filename, 'w', encoding='utf-8') as f:
        f.write(content)

dir_path = '/Users/braulio/Desktop/IDC_ComercialCabello/db-init'
process_file(os.path.join(dir_path, 'init.sql'))
process_file(os.path.join(dir_path, 'dump.sql'))
print('Migration Python OK')

File: db-init/migrate.js
const fs = require('fs');
const path = require('path');

const userMapping = {
  'andrade': 1,
  'arboilito': 2,
  'chosina': 3,
  'crepusculobra': 4,
  'danone': 5,
  'jcamaney': 6,
  'koreanite': 7,
  'miniona': 8,
  'nemo12': 9,
  'ponchalo': 10,
  'rnata': 11
};

const hashToPlain = {
  '$2b$10$n2TX9TtudzDP8lniHpFhXeoa/.nesXPtLwrjr5Zt30AAYNlyUo7KW': '1',
  '$2b$10$nBAFdUOMEK5litSyVEg/UOvcBFdCcxTxp9ctbNQ5eRazTeNDrdwGi': '1234',
  '$2b$10$JHWyY0UiBIN23YE7Vn8wkOqheV.iSCpvvX38HvYwSsZ7AFnyQun/S': '123',
  '$2b$10$vH/3cNMiAl5fpK8zjFfvPO6V1naNqO33TQ4.yJz5m.q7ZfSq5MgKC': '123',
  '$2b$10$cpRu/Lx561rVy6EQft2i7OW8bnPZLIG2xPYyPc6piFPBxK9gh07Py': '1234',
  '$2b$10$e3ltDmY4jX0UWGU.rr/HF.NCkZwqcVr1QjOAIpzpNC93iJ4/8ZzUS': '1234',
  '$2b$10$lyEQpSRipZBmN3OBNEPWa.e9h8ToyLRF4gSXq3qVWYyZ7a5JMiBBy': 'miamors',
  '$2b$10$PX9g.GBUo6RoOb/EwuLnOO6kvtPt8yMhsO.AgR66V8xNe1EjgMOPC': '1234',
  '$2b$10$sFcQ4YGOaIGWio9XRYWD4eqdZ/h4b9CkIpawcsHzp0xun6EXWyOoa': 'renata',
  '$2b$10$bE.ceZevG5jcHv09fXXPxOSGkT6lN7Cq1f7yY0QXnil6pwSDIEmHm': '123',
  '$2b$10$EG5FCjIwX6eqgMlsQ.Tzie47IaPTyIqA8coX4cOqp2yxO0H6qfmxe': 'llanas'
};

function processSql(filename) {
    let content = fs.readFileSync(filename, 'utf8');

    // 1. Remove contrasenas table and inserts
    const dropContrasenas = /DROP TABLE IF EXISTS `contrasenas`;\s*.*\s*.*\s*CREATE TABLE `contrasenas` \([\s\S]*?ENGINE=InnoDB.*?;[\s\S]*?UNLOCK TABLES;/g;
    content = content.replace(dropContrasenas, '');

    // 2. Modify Trabajadores table structure
    const createTrabajadoresRegex = /CREATE TABLE `Trabajadores` \(\s*`usuario` varchar\(20\) NOT NULL,\s*`contrasena` varchar\(128\) NOT NULL,\s*`rol` varchar\(15\) NOT NULL,\s*`Nombre` varchar\(30\) NOT NULL,\s*`apellido_paterno` varchar\(20\) NOT NULL,\s*`apellido_materno` varchar\(20\) NOT NULL,\s*PRIMARY KEY \(`usuario`\)\s*\) ENGINE=InnoDB/g;
    
    const newTrabajadores = `CREATE TABLE \`Trabajadores\` (
  \`id\` int NOT NULL AUTO_INCREMENT,
  \`usuario\` varchar(20) NOT NULL,
  \`contrasena\` varchar(128) NOT NULL,
  \`texto_plano\` varchar(25) DEFAULT NULL,
  \`rol\` varchar(15) NOT NULL,
  \`Nombre\` varchar(30) NOT NULL,
  \`apellido_paterno\` varchar(20) NOT NULL,
  \`apellido_materno\` varchar(20) NOT NULL,
  PRIMARY KEY (\`id\`),
  UNIQUE KEY \`usuario\` (\`usuario\`)
) ENGINE=InnoDB AUTO_INCREMENT=20`;
    content = content.replace(createTrabajadoresRegex, newTrabajadores);

    // 3. Process Trabajadores INSERT
    const insertTrabajadoresRegex = /INSERT INTO `Trabajadores` VALUES (.*);/g;
    content = content.replace(insertTrabajadoresRegex, (match, p1) => {
        // extract the array of values inside parentheses: (...) , (...)
        let chunks = p1.split(/,\s*\(/);
        chunks = chunks.map(chunk => chunk.startsWith('(') ? chunk : '(' + chunk);
        
        let newChunks = chunks.map(chunk => {
            // chunk is like ('andrade','$2b$10$xxx','Operario','Andrade','Chehue','Hernández')
            // we need to insert id and texto_plano.
            const parts = chunk.match(/('[^']*')/g);
            if(parts && parts.length === 6) {
                const usuario = parts[0].replace(/'/g, "");
                const hash = parts[1].replace(/'/g, "");
                const id = userMapping[usuario] || 99;
                const plain = hashToPlain[hash] || '';
                
                return `(${id},${parts[0]},${parts[1]},'${plain}',${parts[2]},${parts[3]},${parts[4]},${parts[5]})`;
            }
            return chunk;
        });

        return `INSERT INTO \`Trabajadores\` VALUES ${newChunks.join(',')};`;
    });

    // 4. Modify ventas table foreign key logic
    const createVentasRegex = /CREATE TABLE `ventas` \([\s\S]*?`usuario` varchar\(20\) DEFAULT NULL,[\s\S]*?KEY `usuario` \(`usuario`\),[\s\S]*?CONSTRAINT `ventas_ibfk_2` FOREIGN KEY \(`usuario`\) REFERENCES `trabajadores` \(`usuario`\)[\s\S]*?\) ENGINE=InnoDB/;
    
    // Instead of completely matching the big chunk, let's just replace column and constraint:
    content = content.replace(/`usuario` varchar\(20\) DEFAULT NULL,/g, '`trabajador_id` int DEFAULT NULL,');
    content = content.replace(/KEY `usuario` \(`usuario`\)/g, 'KEY `trabajador_id` (`trabajador_id`)');
    content = content.replace(/CONSTRAINT `ventas_ibfk_2` FOREIGN KEY \(`usuario`\) REFERENCES `trabajadores` \(`usuario`\)/gi, 'CONSTRAINT `ventas_ibfk_2` FOREIGN KEY (`trabajador_id`) REFERENCES `Trabajadores` (`id`)');

    // 5. Update Ventas INSERTs where users are strings 'danone', 'miniona', etc...
    const insertVentasRegex = /INSERT INTO `ventas` VALUES (.*);/g;
    content = content.replace(insertVentasRegex, (match, p1) => {
        let chunks = p1.split(/,\s*\(/);
        chunks = chunks.map(chunk => chunk.startsWith('(') ? chunk : '(' + chunk);
        
        let newChunks = chunks.map(chunk => {
            // chunk e.g. (24,1,'2839843490393',1,27,'2025-05-28','danone')
            // last part is the user string
            const parts = chunk.split(',');
            const lastPart = parts[parts.length - 1]; // e.g. 'danone')
            const userMatch = lastPart.match(/'([^']+)'\)/);
            if(userMatch) {
                const user = userMatch[1];
                if(userMapping[user]) {
                    parts[parts.length - 1] = `${userMapping[user]})`;
                }
            }
            return parts.join(',');
        });
        return `INSERT INTO \`ventas\` VALUES ${newChunks.join(',')};`;
    });

    fs.writeFileSync(filename, content);
}

const dir = '/Users/braulio/Desktop/IDC_ComercialCabello/db-init';
processSql(path.join(dir, 'init.sql'));
processSql(path.join(dir, 'dump.sql'));
console.log('Migration OK');
